File: platform/packages_py/auth_encoding/auth_encoding/fetch_auth_encoding.py
"""
Core authentication encoding functionality.
Encodes credentials into HTTP headers for various auth types.
"""
import base64
from typing import Dict, Any

from .types.auth_type import AuthType
from .types.credentials import (
    USERNAME_KEYS,
    PASSWORD_KEYS,
    EMAIL_KEYS,
    TOKEN_KEYS,
    API_KEY_KEYS,
    HEADER_KEY_KEYS,
    HEADER_VALUE_KEYS,
    extract_credential,
)
from .errors import (
    MissingCredentialError,
    InvalidAuthTypeError,
    HMACNotImplementedError,
)


def _to_base64(s: str) -> str:
    """Encode a string to Base64."""
    return base64.b64encode(s.encode("utf-8")).decode("utf-8")


def _encode_basic_auth(username: str, password: str) -> str:
    """Encode basic authentication (username:password) to Base64."""
    return _to_base64(f"{username}:{password}")


def encode_auth(auth_type: str, credentials: Dict[str, Any]) -> Dict[str, str]:
    """
    Encode authentication credentials into HTTP headers.

    Args:
        auth_type: The authentication type
        credentials: The credentials to encode

    Returns:
        Dictionary of header name to header value

    Raises:
        MissingCredentialError: If required credentials are missing
        InvalidAuthTypeError: If auth type is not supported
        HMACNotImplementedError: If HMAC auth is attempted
    """
    normalized_type = auth_type.lower()

    # Basic authentication types
    if normalized_type == AuthType.BASIC:
        username = extract_credential(credentials, USERNAME_KEYS)
        password = extract_credential(credentials, PASSWORD_KEYS)
        if not username:
            raise MissingCredentialError(normalized_type, "username")
        if not password:
            raise MissingCredentialError(normalized_type, "****")
        return {"Authorization": f"Basic {_encode_basic_auth(username, password)}"}

    elif normalized_type == AuthType.BASIC_EMAIL:
        email = extract_credential(credentials, EMAIL_KEYS)
        password = extract_credential(credentials, PASSWORD_KEYS)
        if not email:
            raise MissingCredentialError(normalized_type, "email")
        if not password:
            raise MissingCredentialError(normalized_type, "****")
        return {"Authorization": f"Basic {_encode_basic_auth(email, password)}"}

    elif normalized_type == AuthType.BASIC_TOKEN:
        username = extract_credential(credentials, USERNAME_KEYS)
        token = extract_credential(credentials, TOKEN_KEYS)
        if not username:
            raise MissingCredentialError(normalized_type, "username")
        if not token:
            raise MissingCredentialError(normalized_type, "token")
        return {"Authorization": f"Basic {_encode_basic_auth(username, token)}"}

    elif normalized_type == AuthType.BASIC_EMAIL_TOKEN:
        email = extract_credential(credentials, EMAIL_KEYS)
        token = extract_credential(credentials, TOKEN_KEYS)
        if not email:
            raise MissingCredentialError(normalized_type, "email")
        if not token:
            raise MissingCredentialError(normalized_type, "token")
        return {"Authorization": f"Basic {_encode_basic_auth(email, token)}"}

    # Bearer token types - raw token
    elif normalized_type in (AuthType.BEARER, AuthType.BEARER_OAUTH, AuthType.BEARER_JWT):
        token = extract_credential(credentials, TOKEN_KEYS)
        if not token:
            raise MissingCredentialError(normalized_type, "token")
        return {"Authorization": f"Bearer {token}"}

    # Bearer token types - compound (Base64 encoded)
    elif normalized_type == AuthType.BEARER_USERNAME_TOKEN:
        username = extract_credential(credentials, USERNAME_KEYS)
        token = extract_credential(credentials, TOKEN_KEYS)
        if not username:
            raise MissingCredentialError(normalized_type, "username")
        if not token:
            raise MissingCredentialError(normalized_type, "token")
        return {"Authorization": f"Bearer {_encode_basic_auth(username, token)}"}

    elif normalized_type == AuthType.BEARER_USERNAME_PASSWORD:
        username = extract_credential(credentials, USERNAME_KEYS)
        password = extract_credential(credentials, PASSWORD_KEYS)
        if not username:
            raise MissingCredentialError(normalized_type, "username")
        if not password:
            raise MissingCredentialError(normalized_type, "****")
        return {"Authorization": f"Bearer {_encode_basic_auth(username, password)}"}

    elif normalized_type == AuthType.BEARER_EMAIL_TOKEN:
        email = extract_credential(credentials, EMAIL_KEYS)
        token = extract_credential(credentials, TOKEN_KEYS)
        if not email:
            raise MissingCredentialError(normalized_type, "email")
        if not token:
            raise MissingCredentialError(normalized_type, "token")
        return {"Authorization": f"Bearer {_encode_basic_auth(email, token)}"}

    elif normalized_type == AuthType.BEARER_EMAIL_PASSWORD:
        email = extract_credential(credentials, EMAIL_KEYS)
        password = extract_credential(credentials, PASSWORD_KEYS)
        if not email:
            raise MissingCredentialError(normalized_type, "email")
        if not password:
            raise MissingCredentialError(normalized_type, "****")
        return {"Authorization": f"Bearer {_encode_basic_auth(email, password)}"}

    # API key authentication
    elif normalized_type == AuthType.X_API_KEY:
        api_key = extract_credential(credentials, API_KEY_KEYS)
        if not api_key:
            raise MissingCredentialError(normalized_type, "apiKey")
        return {"X-API-Key": api_key}

    # Custom header authentication
    elif normalized_type in (AuthType.CUSTOM, AuthType.CUSTOM_HEADER):
        header_key = extract_credential(credentials, HEADER_KEY_KEYS)
        header_value = extract_credential(credentials, HEADER_VALUE_KEYS)
        if not header_key:
            raise MissingCredentialError(normalized_type, "headerKey")
        if not header_value:
            raise MissingCredentialError(normalized_type, "headerValue")
        return {header_key: header_value}

    # No authentication
    elif normalized_type == AuthType.NONE:
        return {}

    # HMAC - reserved but not implemented
    elif normalized_type == AuthType.HMAC:
        raise HMACNotImplementedError()

    else:
        raise InvalidAuthTypeError(auth_type)


def get_header_name(auth_type: str) -> str:
    """Get header name for an auth type."""
    normalized_type = auth_type.lower()

    if normalized_type == AuthType.X_API_KEY:
        return "X-API-Key"
    elif normalized_type in (AuthType.CUSTOM, AuthType.CUSTOM_HEADER):
        return "custom"
    elif normalized_type == AuthType.NONE:
        return "none"
    else:
        return "Authorization"


def is_encoded_auth_type(auth_type: str) -> bool:
    """Check if an auth type uses Base64 encoding."""
    normalized_type = auth_type.lower()

    encoded_types = (
        AuthType.BASIC,
        AuthType.BASIC_EMAIL,
        AuthType.BASIC_TOKEN,
        AuthType.BASIC_EMAIL_TOKEN,
        AuthType.BEARER_USERNAME_TOKEN,
        AuthType.BEARER_USERNAME_PASSWORD,
        AuthType.BEARER_EMAIL_TOKEN,
        AuthType.BEARER_EMAIL_PASSWORD,
    )

    return normalized_type in encoded_types

File: platform/packages_mjs/auth_encoding/src/fetch-auth-encoding.ts
/**
 * Core authentication encoding functionality.
 * Encodes credentials into HTTP headers for various auth types.
 */

import { AuthType } from './types/auth-type.js';
import {
    type AuthCredentials,
    USERNAME_KEYS,
    PASSWORD_KEYS,
    EMAIL_KEYS,
    TOKEN_KEYS,
    API_KEY_KEYS,
    HEADER_KEY_KEYS,
    HEADER_VALUE_KEYS,
    extractCredential,
} from './types/credentials.js';
import {
    MissingCredentialError,
    InvalidAuthTypeError,
    HMACNotImplementedError,
} from './errors.js';

/**
 * Encode a string to Base64
 */
function toBase64(str: string): string {
    return Buffer.from(str, 'utf-8').toString('base64');
}

/**
 * Encode basic authentication (username:password) to Base64
 */
function encodeBasicAuth(username: string, password: string): string {
    return toBase64(`${username}:${password}`);
}

/**
 * Encode authentication credentials into HTTP headers.
 *
 * @param authType - The authentication type
 * @param credentials - The credentials to encode
 * @returns Record of header name to header value
 * @throws MissingCredentialError if required credentials are missing
 * @throws InvalidAuthTypeError if auth type is not supported
 * @throws HMACNotImplementedError if HMAC auth is attempted
 */
export function encodeAuth(
    authType: string,
    credentials: AuthCredentials
): Record<string, string> {
    const normalizedType = authType.toLowerCase();

    switch (normalizedType) {
        // Basic authentication types
        case AuthType.BASIC: {
            const username = extractCredential(credentials, USERNAME_KEYS);
            const password = extractCredential(credentials, PASSWORD_KEYS);
            if (!username) {
                throw new MissingCredentialError(normalizedType, 'username');
            }
            if (!password) {
                throw new MissingCredentialError(normalizedType, 'password');
            }
            return {
                Authorization: `Basic ${encodeBasicAuth(username, password)}`,
            };
        }

        case AuthType.BASIC_EMAIL: {
            const email = extractCredential(credentials, EMAIL_KEYS);
            const password = extractCredential(credentials, PASSWORD_KEYS);
            if (!email) {
                throw new MissingCredentialError(normalizedType, 'email');
            }
            if (!password) {
                throw new MissingCredentialError(normalizedType, 'password');
            }
            return {
                Authorization: `Basic ${encodeBasicAuth(email, password)}`,
            };
        }

        case AuthType.BASIC_TOKEN: {
            const username = extractCredential(credentials, USERNAME_KEYS);
            const token = extractCredential(credentials, TOKEN_KEYS);
            if (!username) {
                throw new MissingCredentialError(normalizedType, 'username');
            }
            if (!token) {
                throw new MissingCredentialError(normalizedType, 'token');
            }
            return {
                Authorization: `Basic ${encodeBasicAuth(username, token)}`,
            };
        }

        case AuthType.BASIC_EMAIL_TOKEN: {
            const email = extractCredential(credentials, EMAIL_KEYS);
            const token = extractCredential(credentials, TOKEN_KEYS);
            if (!email) {
                throw new MissingCredentialError(normalizedType, 'email');
            }
            if (!token) {
                throw new MissingCredentialError(normalizedType, 'token');
            }
            return {
                Authorization: `Basic ${encodeBasicAuth(email, token)}`,
            };
        }

        // Bearer token types - raw token
        case AuthType.BEARER:
        case AuthType.BEARER_OAUTH:
        case AuthType.BEARER_JWT: {
            const token = extractCredential(credentials, TOKEN_KEYS);
            if (!token) {
                throw new MissingCredentialError(normalizedType, 'token');
            }
            return {
                Authorization: `Bearer ${token}`,
            };
        }

        // Bearer token types - compound (Base64 encoded)
        case AuthType.BEARER_USERNAME_TOKEN: {
            const username = extractCredential(credentials, USERNAME_KEYS);
            const token = extractCredential(credentials, TOKEN_KEYS);
            if (!username) {
                throw new MissingCredentialError(normalizedType, 'username');
            }
            if (!token) {
                throw new MissingCredentialError(normalizedType, 'token');
            }
            return {
                Authorization: `Bearer ${encodeBasicAuth(username, token)}`,
            };
        }

        case AuthType.BEARER_USERNAME_PASSWORD: {
            const username = extractCredential(credentials, USERNAME_KEYS);
            const password = extractCredential(credentials, PASSWORD_KEYS);
            if (!username) {
                throw new MissingCredentialError(normalizedType, 'username');
            }
            if (!password) {
                throw new MissingCredentialError(normalizedType, 'password');
            }
            return {
                Authorization: `Bearer ${encodeBasicAuth(username, password)}`,
            };
        }

        case AuthType.BEARER_EMAIL_TOKEN: {
            const email = extractCredential(credentials, EMAIL_KEYS);
            const token = extractCredential(credentials, TOKEN_KEYS);
            if (!email) {
                throw new MissingCredentialError(normalizedType, 'email');
            }
            if (!token) {
                throw new MissingCredentialError(normalizedType, 'token');
            }
            return {
                Authorization: `Bearer ${encodeBasicAuth(email, token)}`,
            };
        }

        case AuthType.BEARER_EMAIL_PASSWORD: {
            const email = extractCredential(credentials, EMAIL_KEYS);
            const password = extractCredential(credentials, PASSWORD_KEYS);
            if (!email) {
                throw new MissingCredentialError(normalizedType, 'email');
            }
            if (!password) {
                throw new MissingCredentialError(normalizedType, 'password');
            }
            return {
                Authorization: `Bearer ${encodeBasicAuth(email, password)}`,
            };
        }

        // API key authentication
        case AuthType.X_API_KEY: {
            const apiKey = extractCredential(credentials, API_KEY_KEYS);
            if (!apiKey) {
                throw new MissingCredentialError(normalizedType, 'apiKey');
            }
            return {
                'X-API-Key': apiKey,
            };
        }

        // Custom header authentication
        case AuthType.CUSTOM:
        case AuthType.CUSTOM_HEADER: {
            const headerKey = extractCredential(credentials, HEADER_KEY_KEYS);
            const headerValue = extractCredential(credentials, HEADER_VALUE_KEYS);
            if (!headerKey) {
                throw new MissingCredentialError(normalizedType, 'headerKey');
            }
            if (!headerValue) {
                throw new MissingCredentialError(normalizedType, 'headerValue');
            }
            return {
                [headerKey]: headerValue,
            };
        }

        // No authentication
        case AuthType.NONE: {
            return {};
        }

        // HMAC - reserved but not implemented
        case AuthType.HMAC: {
            throw new HMACNotImplementedError();
        }

        default: {
            throw new InvalidAuthTypeError(authType);
        }
    }
}

/**
 * Get header name for an auth type
 */
export function getHeaderName(authType: string): string {
    const normalizedType = authType.toLowerCase();

    switch (normalizedType) {
        case AuthType.X_API_KEY:
            return 'X-API-Key';
        case AuthType.CUSTOM:
        case AuthType.CUSTOM_HEADER:
            return 'custom';
        case AuthType.NONE:
            return 'none';
        default:
            return 'Authorization';
    }
}

/**
 * Check if an auth type uses Base64 encoding
 */
export function isEncodedAuthType(authType: string): boolean {
    const normalizedType = authType.toLowerCase();

    const encodedTypes: string[] = [
        AuthType.BASIC,
        AuthType.BASIC_EMAIL,
        AuthType.BASIC_TOKEN,
        AuthType.BASIC_EMAIL_TOKEN,
        AuthType.BEARER_USERNAME_TOKEN,
        AuthType.BEARER_USERNAME_PASSWORD,
        AuthType.BEARER_EMAIL_TOKEN,
        AuthType.BEARER_EMAIL_PASSWORD,
    ];

    return encodedTypes.includes(normalizedType);
}
